<a href="https://colab.research.google.com/github/abdullahb07/ml-models/blob/main/Copy_of_M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [34]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL
data = yf.download("AAPL", start="2019-01-01", end="2023-01-01")

# Display summary statistics
print(data.describe())

/tmp/ipython-input-34-1495478674.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2019-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price         Close         High          Low         Open        Volume
Ticker         AAPL         AAPL         AAPL         AAPL          AAPL
count   1008.000000  1008.000000  1008.000000  1008.000000  1.008000e+03
mean     108.238322   109.494262   106.872874   108.156611  1.120997e+08
std       42.912798    43.456870    42.358244    42.930863  5.310973e+07
min       33.870827    34.711701    33.825567    34.297218  3.519590e+07
25%       64.530735    65.189512    63.730465    64.412146  7.755390e+07
50%      120.808475   122.041300   119.235587   120.879095  9.693830e+07
75%      144.630890   146.087770   142.759991   144.422373  1.290616e+08
max      178.645630   179.558473   175.809076   179.254206  4.265100e+08


In [12]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())

/tmp/ipython-input-12-1526985050.py:9: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633102  129.181519  125.971433  127.555152  1.120920e+08
std      30.421463   30.698899   30.139550   30.449563  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709515  114.204952  111.251831  112.889666  7.636470e+07
50%     133.010773  134.143863  131.050657  132.832322  9.493580e+07
75%     148.221500  149.388580  146.276911  147.588786  1.296327e+08
max     178.645615  179.558442  175.809046  179.254175  4.265100e+08


In [13]:
# Prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01
# using Yahoo Finance and summarize its key statistics.

import yfinance as yf

data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())

/tmp/ipython-input-13-2222841701.py:6: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")
[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.633102  129.181519  125.971433  127.555152  1.120920e+08
std      30.421463   30.698899   30.139550   30.449563  5.602586e+07
min      54.378582   55.379535   51.528416   55.277744  3.519590e+07
25%     112.709515  114.204952  111.251831  112.889666  7.636470e+07
50%     133.010773  134.143863  131.050657  132.832322  9.493580e+07
75%     148.221500  149.388580  146.276911  147.588786  1.296327e+08
max     178.645615  179.558442  175.809046  179.254175  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



In [9]:
# Feature Engineering: Add historical prices, volume, and moving averages

data['Open_Lag1'] = data['Open'].shift(1)
data['High_Lag1'] = data['High'].shift(1)
data['Low_Lag1'] = data['Low'].shift(1)
data['Close_Lag1'] = data['Close'].shift(1)
data['Volume_Lag1'] = data['Volume'].shift(1)

data['MA_5'] = data['Close'].rolling(window=5).mean()
data['MA_10'] = data['Close'].rolling(window=10).mean()
data['MA_50'] = data['Close'].rolling(window=50).mean()

# Drop rows with NaN values created by shifting and rolling
data.dropna(inplace=True)

# Define features (X) and target (y)
features = ['Open_Lag1', 'High_Lag1', 'Low_Lag1', 'Close_Lag1', 'Volume_Lag1', 'MA_5', 'MA_10', 'MA_50']
target = 'Close' # Predict tomorrow's closing price

X = data[features]
y = data[target]

# Split data into training and testing sets (70/30 ratio)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, shuffle=False)

print("Training data shape:", X_train.shape, y_train.shape)
print("Testing data shape:", X_test.shape, y_test.shape)

Training data shape: (494, 8) (494, 1)
Testing data shape: (213, 8) (213, 1)


In [10]:
# Train a Linear Regression model

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R²) Score: {r2:.2f}")

Mean Squared Error (MSE): 8.44
R-squared (R²) Score: 0.94


In [29]:
# Remove this code as it is causing errors

In [30]:
# Remove this code as it is causing errors

In [35]:
# Create a binary target variable (1 if close price increases tomorrow, 0 otherwise)
data['Price_Increase'] = (data['Close'].shift(-1) > data['Close']).astype(int)

# Drop the last row as it will have NaN for the target variable
data.dropna(inplace=True)

# Define features (X) and target (y)
# Using the same features as the regression model
features = ['Open', 'High', 'Low', 'Close', 'Volume'] # Using basic historical features for simplicity first
target = 'Price_Increase'

X = data[features]
y = data[target]

# Split data into training and testing sets (70/30 ratio)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, shuffle=False)

print("Training data shape:", X_train.shape, y_train.shape)
print("Testing data shape:", X_test.shape, y_test.shape)

Training data shape: (705, 5) (705,)
Testing data shape: (303, 5) (303,)


In [36]:
# Train a Random Forest Classifier model

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

model_rf = RandomForestClassifier(n_estimators=100, random_state=42)
model_rf.fit(X_train, y_train)

# Predict on the test set
y_pred_rf = model_rf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred_rf)
report = classification_report(y_test, y_pred_rf)

print(f"Accuracy: {accuracy:.2f}")
print("Classification Report:")
print(report)

# Report feature importance
feature_importance = pd.Series(model_rf.feature_importances_, index=features).sort_values(ascending=False)
print("\nFeature Importance:")
print(feature_importance)

Accuracy: 0.54
Classification Report:
              precision    recall  f1-score   support

           0       0.54      0.79      0.64       156
           1       0.56      0.29      0.38       147

    accuracy                           0.54       303
   macro avg       0.55      0.54      0.51       303
weighted avg       0.55      0.54      0.51       303


Feature Importance:
Volume    0.229240
High      0.197935
Close     0.196932
Open      0.192445
Low       0.183448
dtype: float64


In [39]:
# Download historical data for AAPL for the prediction period (2023-01-01 to 2024-01-01)
prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")

# Feature Engineering for the prediction data - apply the same features as used for training
# Using only the basic historical features as the model was trained on these
features = ['Open', 'High', 'Low', 'Close', 'Volume']

# Create the actual target variable for the prediction period
prediction_data['Price_Increase'] = (prediction_data['Close'].shift(-1) > prediction_data['Close']).astype(int)

# Drop the last row as it will have NaN for the target variable
prediction_data.dropna(inplace=True)

# Define features (X_predict) and actual target (y_predict) for the prediction period
target = 'Price_Increase'

X_predict = prediction_data[features]
y_predict = prediction_data[target]

print("Prediction data shape:", X_predict.shape, y_predict.shape)

/tmp/ipython-input-39-1933691476.py:2: FutureWarning: YF.download() has changed argument auto_adjust default to True
  prediction_data = yf.download("AAPL", start="2023-01-01", end="2024-01-01")
[*********************100%***********************]  1 of 1 completed

Prediction data shape: (250, 5) (250,)


In [40]:
# Make predictions on the prediction data
y_pred_predict = model_rf.predict(X_predict)

# Evaluate the model performance on the prediction data
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

accuracy_predict = accuracy_score(y_predict, y_pred_predict)
report_predict = classification_report(y_predict, y_pred_predict)
confusion_matrix_predict = confusion_matrix(y_predict, y_pred_predict)

print(f"Accuracy on prediction data: {accuracy_predict:.2f}")
print("\nConfusion Matrix on prediction data:")
print(confusion_matrix_predict)
print("\nClassification Report on prediction data:")
print(report_predict)

Accuracy on prediction data: 0.50

Confusion Matrix on prediction data:
[[102   8]
 [117  23]]

Classification Report on prediction data:
              precision    recall  f1-score   support

           0       0.47      0.93      0.62       110
           1       0.74      0.16      0.27       140

    accuracy                           0.50       250
   macro avg       0.60      0.55      0.44       250
weighted avg       0.62      0.50      0.42       250



## Model Performance and Trading Strategy Discussion

Based on the evaluation of the Random Forest classification model on the prediction data from 2023-01-01 to 2024-01-01:

**Model Performance:**

The model achieved an overall accuracy of approximately 50%, which is close to random chance. The classification report revealed that the model is better at identifying days when the stock price is *not* likely to increase (high recall for class 0: 0.93) but struggles significantly to predict days when the price *will* increase (low recall for class 1: 0.16). While the precision for class 1 is higher (0.74), the low recall means it rarely makes a correct positive prediction.

In summary, the model is not reliable for predicting upward price movements and its overall performance is not sufficient for making confident trading decisions.

**Potential Trading Strategies (with significant caution):**

Given the model's limitations, any trading strategy based on its predictions would be highly risky. Theoretical strategies, acknowledging these limitations, could include:

1.  **Conservative "Avoid Downside" Strategy:** Focus on using the model's relatively higher recall for predicting no price increase (class 0) to potentially avoid buying or consider selling on those days. This aims to mitigate losses but would result in missed opportunities due to the lower precision for this class.

2.  **Extremely Cautious "High Precision Buy" Strategy:** A highly speculative approach could be to only consider buying when the model predicts a price increase (class 1) with a very high probability threshold. This strategy would yield very few trades due to the low recall for class 1 and still carries substantial risk.

**Critical Considerations:**

This analysis reinforces the difficulty of stock price prediction. The simple features and model used are not adequate to capture market complexities. Relying on a model with this performance for real-world trading is not advisable and could lead to significant financial losses. Improving the model would require exploring more advanced techniques, incorporating a broader range of features (including external economic and market indicators), and conducting rigorous backtesting and validation.